data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,392 @@ category: "Module System"
+
+Here is the skeleton of a possible implementation:
+
+**heap.ml**
+```ocaml
+module type OrderedType = sig
+  type t
+  val compare : t -> t -> int
+end
+
+module type S = sig
+  type elt
+  type t
+  val empty : t
+  val is_empty : t -> bool
+  val insert : t -> elt -> t
+  val merge : t -> t -> t
+  val find : t -> elt
+  val delete : t -> t
+end
+
+module Binary(Elt: OrderedType) : S = struct
+  type elt = | (* Replace by your own *)
+  type t = | (* Replace by your own *)
+  (* Add private functions here *)
+  let is_empty h = failwith "Not yet implemented"
+  let insert h e = failwith "Not yet implemented"
+  let merge h1 h2 = failwith "Not yet implemented"
+  let find h = failwith "Not yet implemented"
+  let delete h = failwith "Not yet implemented"
+end
+```
+
+Here, binary heaps is the only implementation suggested. This can be extended to other implementations by adding one functor per each (e.g., `Heap.Leftist`, `Heap.Binomial`, `Heap.Fibonacci`, etc.).
+
+## Injecting Dependencies Using Functors
+
+**Dependencies Between Modules**
+
+Here is a new version of the `funkt` program:
+
+**`funkt.ml`**
+```ocaml
+module StringSet = Set.Make(String)
+
+module IterPrint : sig
+  val f : string list -> unit
+end = struct
+  let f = List.iter (fun s -> Out_channel.output_string stdout (s ^ "\n"))
+end
+
+let _ =
+  stdin
+  |> In_channel.input_lines
+  |> List.concat_map Str.(split (regexp "[ \t.,;:()]+"))
+  |> StringSet.of_list
+  |> StringSet.elements
+  |> IterPrint.f
+```
+
+It embeds an additional `IterPrint` module that exposes a single function `f` of type `string list -> unit` and has two dependencies:
+  - Module `List` through `List.iter` and the type of its `f` function
+  - Module `Out_channel` through `Out_channel.output_string`
+
+Check the behaviour of the program using `dune exec funkt < dune`.
+
+**Dependency Injection**
+
+[Dependency injection](https://en.wikipedia.org/wiki/Dependency_injection) is a way to parametrise over a dependency.
+
+Here is a refactoring of the module `IterPrint` to make of this technique:
+
+**`iterPrint.ml`**
+```ocaml
+module type Iterable = sig
+  type 'a t
+  val iter : ('a -> unit) -> 'a t -> unit
+end
+
+module type S = sig
+  type 'a t
+  val f : string t -> unit
+end
+
+module Make(Dep: Iterable) : S with type 'a t := 'a Dep.t = struct
+  let f = Dep.iter (fun s -> Out_channel.output_string stdout (s ^ "\n"))
+end
+```
+
+The module `IterPrint` is refactored into a functor that takes the dependency providing `iter` as a parameter. The `with type 'a t := 'a Dep.t` constraint means the type `t` from the parameter `Dep` replaces the type `t` in the result module. This allows the type of `f` to use the type `t` from the parameter module `Dep`. With this refactoring, `IterPrint` only has one dependency; at the time it is compiled, no implementation of function `iter` is available yet.
```

Comment:
```suggestion
The module `IterPrint` is refactored into a functor that takes the dependency providing the parameter `iter`. The `with type 'a t := 'a Dep.t` constraint means the type `t` from the parameter `Dep` replaces the type `t` in the result module. This allows the type of `f` to use the type `t` from the parameter module `Dep`. With this refactoring, `IterPrint` only has one dependency. At the time it is compiled, no implementation of function `iter` is available yet.
```

If this isn't accurate, let's work on the syntax together.

Reply:
Check updated text line 312

Reply:
Much better! I suggested two little tweaks.